lib/scholar/impute/simple_imputer.ex

```diff
@@ -0,0 +1,215 @@
+        else: x
+
+    {_num_rows, num_cols} = Nx.shape(x)
+
+    statistics =
+      cond do
+        opts[:strategy] == :mean ->
+          mean_op(x)
+
+        opts[:strategy] == :median ->
+          median_op(x)
+
+        opts[:strategy] == :mode ->
+          mode_op(x, type: x_type)
+
+        true ->
+          Nx.broadcast(validated_opts[:fill_value], {num_cols})
+      end
+
+    missing_values = validated_opts[:missing_values]
+    %__MODULE__{statistics: statistics, missing_values: missing_values}
+  end
+
+  defnp mean_op(x) do
+    mask = not Nx.is_nan(x)
+    denominator = Nx.sum(mask, axes: [0])
+    temp = Nx.select(mask, x, 0)
+    nominator = Nx.sum(temp, axes: [0])
+
+    Nx.select(
+      denominator != 0,
+      nominator / denominator,
+      Nx.Constants.nan()
+    )
+  end
+
+  defnp median_op(x) do
+    {num_rows, num_cols} = Nx.shape(x)
+    x = Nx.sort(x, axis: 0)
+    tensor = Nx.iota(x, axis: 0) * num_rows
+    res = Nx.argsort(tensor, axis: 0)
+    x = Nx.take_along_axis(x, res, axis: 0)
+
+    indices = Nx.broadcast(num_rows - 1, {num_cols})
+    indices = indices - Nx.sum(Nx.is_nan(x), axes: [0])
+    half_indices = indices / 2
+    floor = Nx.as_type(Nx.floor(half_indices), :s64) |> Nx.new_axis(0)
+    ceil = Nx.as_type(Nx.ceil(half_indices), :s64) |> Nx.new_axis(0)
+    nums1 = Nx.take_along_axis(x, floor, axis: 0)
+    nums2 = Nx.take_along_axis(x, ceil, axis: 0)
+    ((nums1 + nums2) / 2) |> Nx.squeeze()
+  end
+
+  defnp mode_op(x, opts \\ []) do
+    type = opts[:type]
+
+    axis = 0
+
+    axis_length = Nx.axis_size(x, axis)
+    x = Nx.sort(x, axis: 0)
+    tensor = Nx.multiply(Nx.iota(x, axis: axis), axis_length)
+    res = Nx.argsort(tensor, axis: axis)
+    sorted = Nx.take_along_axis(x, res, axis: axis)
+
+    {num_rows, num_cols} = Nx.shape(tensor)
+
+    num_elements = Nx.size(tensor)
+
+    group_indices =
+      Nx.concatenate(
+        [
+          Nx.broadcast(0, {1, num_cols}),
+          Nx.not_equal(
+            Nx.slice_along_axis(sorted, 0, num_rows - 1, axis: axis),
+            Nx.slice_along_axis(sorted, 1, num_rows - 1, axis: axis)
+          )
+        ],
+        axis: axis
+      )
+      |> Nx.cumulative_sum(axis: axis)
+
+    counting_indices =
+      [
+        Nx.reshape(group_indices, {num_elements, 1}),
+        group_indices
+        |> Nx.iota(axis: 1)
+        |> Nx.reshape({num_elements, 1})
+      ]
+      |> Nx.concatenate(axis: 1)
+
+    largest_group_indices =
+      Nx.broadcast(0, sorted)
+      |> Nx.indexed_add(counting_indices, Nx.broadcast(1, {num_elements}))
+      |> Nx.argmax(axis: axis, keep_axis: true)
+
+    indices =
+      largest_group_indices
+      |> Nx.broadcast(Nx.shape(group_indices))
+      |> Nx.equal(group_indices)
+      |> Nx.argmax(axis: axis, keep_axis: true)
+
+    Nx.take_along_axis(sorted, indices, axis: axis) |> Nx.squeeze() |> Nx.as_type(type)
+  end
```

Comment:
Should we update the Nx dependency so we can use the mode and median functions in Nx?

Reply:
No, this functions are implemented a little bit different than these in Nx.